### PR TITLE
Fixed bug with updating TextNode in IE8 (#7824)

### LINKED ISSUE #7824: React fails to update TextNode in IE8
**Do you want to request a *feature* or report a *bug*?**

Bug

**What is the current behavior?**

React fails to update text nodes in IE8

**If the current behavior is a bug, please provide the steps to reproduce and if possible a minimal demo of the problem via https://jsfiddle.net or similar (template: https://jsfiddle.net/reactjs/69z2wepo/).**

This is as close as I could get to a reproduction: https://fiddle.jshell.net/zrozubsb/1/show/

I think JSFiddle itself doesn't work in IE8 though, so it's masking the real problem, which you can see here:

image

**What is the expected behavior?**

DOM updates/doesn't throw error

**Which versions of React, and which browser / OS are affected by this issue? Did this work in previous versions of React?**

React 15.3.1. Internet Explorer 8. Not sure if it ever worked; I have a feeling it didn't.

-----

I will be submitting a PR for this shortly as I know you aren't actively supporting IE8 any more, but it's actually a pretty simple fix. I hope I outlined the issue well enough.

### PR DESCRIPTION
**Pull request by mnpenner**
Fixes #7824

### CODE REVIEW

**Comment by syranide** [comment]
That's wrong, it will render the HTML as text rather than HTML. But I'm more curious why you're even suggesting this, can you provide a repro where React is doing the wrong thing and trying to render HTML into a text node?

**Comment by syranide** [comment]
It's bothersome to replicate this because you can't run jsfiddles on IE8 (IE8 emulation is not the same as IE8!), so I don't have the time right now. However, because TextNodes doesn't have `innerHTML` in any browser then this should apply equally to all browsers I would assume, however other browsers do allow you to set `innerHTML` (but it just won't show), so it's possible that this is a bug that isn't showing in other browsers.

So this behavior should definitely be investigated, but I'm quite sure that this PR isn't a correct fix.

**Comment by mnpenner** [comment]
Well, I do have a real IE8 VM, I just hate firing it up because it doesn't even have a proper debugger.

I managed to reproduce it properly now.

image

I've hosted a unadulterated copy of the demo [here](http://nucleuslabs.com/Vhq2v6dX4yPrU9JybasIQyh8jr7beT6WILw/).

I had to run all the libs through webpack to get it to run in IE8. Here's what's included:

- core-js/es5
- core-js/es6
- react
- react-dom
- [Hello.js](http://nucleuslabs.com/Vhq2v6dX4yPrU9JybasIQyh8jr7beT6WILw/Hello.js)

I'm not sure why this bug doesn't affect newer browsers -- is this handled further up the callstack for them? Maybe they don't even enter this `setInnerHTML` function.

**Comment by mnpenner** [comment]
I switched back to IE11 emulating IE8 so that I could get a callstack. It actually goes through here first:

```js
var setTextContent = function(node, text) {
  if (text) {
    var firstChild = node.firstChild;
    
    if (firstChild && firstChild === node.lastChild && firstChild.nodeType === 3) {
      firstChild.nodeValue = text;
      return;
    }
  }
  node.textContent = text;
};

if (ExecutionEnvironment.canUseDOM) {
  if (!('textContent' in document.documentElement)) {
    setTextContent = function(node, text) {
      setInnerHTML(node, escapeTextContentForBrowser(text)); // <-- gets called
    };
  }
}
```

You'll note that it only falls back to `setInnerHTML` if `textContent` is not available, which is true of IE8 but not newer browsers.

So, new browsers should use `node.textContent = text` and my patch will only affect browsers without `textContent`.

**Comment by syranide** [comment]
@mnpenner Ah now that makes a lot more sense :) excellent digging. So `setTextContent` should have a special-case for textNodes and set `nodeValue` directly.

PS. Also, you messed up the PR when you pulled/rebased.

**Comment by mnpenner** [comment]
@syranide Yeah...sorry. Totally different workflow at my company. Don't know how this Git + Github stuff works :stuck_out_tongue_closed_eyes: I was trying to [revert my code](http://stackoverflow.com/a/9646323/65387) to facebook/react master so I could try again.

Would we want to put that check in the fallback `setTextContent` function then, since the original is overwritten?

It seems like it should be in the original too, but it never gets hit AFAIK.

**Comment by syranide** [comment]
@mnpenner Absolutely no worries, was just giving the heads up :)

> Would we want to put that check in the fallback setTextContent function then, since the original is overwritten?

Yeah, so the fallback `setTextContent` should check if `node` is a TextNode, if it is then it should set with `nodeValue`, else it should use `setInnerHTML` as it already does. `setInnerHTML` should not be changed because it's not meant to act on TextNodes.

**Comment by syranide** [comment]
As for the "original" `setTextContent` it is not needed there, TextNodes has `textContent` as well. So yeah, you should technically have a branch for `nodeValue` and one for `textContent`, in practice it is not needed.

File: src/renderers/dom/client/utils/setTextContent.js [hunk 1]
```diff
@@ -28,7 +28,7 @@ var setInnerHTML = require('setInnerHTML');
 var setTextContent = function(node, text) {
   if (text) {
     var firstChild = node.firstChild;
-    
+
```
Comment (syranide) [inline src/renderers/dom/client/utils/setTextContent.js]: Nit, but this wasn't really necessary :)

Comment (mnpenner) [inline src/renderers/dom/client/utils/setTextContent.js]: PhpStorm did that :cry: I noticed after I pushed.

**Comment by syranide** [comment]
Looks good to me 👍

cc @jimfb

**Comment by syranide** [comment]
Oh sorry, the PR is still messed up.

**Comment by mnpenner** [comment]
Uhhh...hate to ask, but do you know the Git commands to set my repo back to a "clean" state so I can try again?

**Pull request closed without merging**